rust/ip-packet/src/icmpv4_header_slice_mut.rs

```diff
@@ -0,0 +1,81 @@
+use crate::slice_utils::write_to_offset_unchecked;
+use etherparse::{
+    icmpv4::{TYPE_ECHO_REPLY, TYPE_ECHO_REQUEST},
+    Icmpv4Slice,
+};
+
+pub struct Icmpv4HeaderSliceMut<'a> {
+    slice: &'a mut [u8],
+}
+
+impl<'a> Icmpv4HeaderSliceMut<'a> {
+    /// Creates a new [`Icmpv4HeaderSliceMut`].
+    pub fn from_slice(slice: &'a mut [u8]) -> Result<Self, etherparse::err::LenError> {
+        Icmpv4Slice::from_slice(slice)?;
+
+        Ok(Self { slice })
+    }
+
+    pub fn set_checksum(&mut self, checksum: u16) {
+        // Safety: Slice is at least of length 8 as checked in the ctor.
+        unsafe { write_to_offset_unchecked(self.slice, 2, checksum.to_be_bytes()) };
+    }
+
+    pub fn set_identifier(&mut self, id: u16) {
+        debug_assert!(
+            self.is_echo_request_or_reply(),
+            "ICMP identifier only exists for echo requests and replies"
+        );
+
+        // Safety: Slice is at least of length 8 as checked in the ctor.
+        unsafe { write_to_offset_unchecked(self.slice, 4, id.to_be_bytes()) };
+    }
+
+    pub fn set_sequence(&mut self, seq: u16) {
+        debug_assert!(
+            self.is_echo_request_or_reply(),
+            "ICMP sequence only exists for echo requests and replies"
+        );
+
+        // Safety: Slice is at least of length 8 as checked in the ctor.
+        unsafe { write_to_offset_unchecked(self.slice, 6, seq.to_be_bytes()) };
+    }
+
+    fn is_echo_request_or_reply(&self) -> bool {
+        let ty = self.slice[0];
+
+        ty == TYPE_ECHO_REPLY || ty == TYPE_ECHO_REQUEST
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use etherparse::{Icmpv4Type, PacketBuilder};
+
+    #[test]
+    fn smoke() {
```

Comment:
I think at this point we should test against the network bytes we expect at this point, feels a bit circular otherwise.

Reply:
I mean in general for all this mutating functions.

Reply:
> I think at this point we should test against the network bytes we expect at this point, feels a bit circular otherwise.

Well, we don't define the immutable slice, right? So assuming that `etherparse` is correct means our code is correct? Verifying the network bytes visually (and thus knowing that the test makes sense) is very difficult.